lib/tasks/eu_exit_business_finder.rake

```diff
@@ -0,0 +1,106 @@
+namespace :eu_exit_business_finder do
+  task update_subscriber_list_titles_and_facet_values: [:environment] do
+    subscriber_lists = SubscriberList
+      .where("links::text != '{}'")
+      .where("ARRAY(SELECT json_object_keys(links)) <@ Array[:keys]", keys: [:facet_values])
+    update_facet_values(subscriber_lists)
+  end
+
+  # Iterate through the affected subscriber lists and if we encounter any
+  # facet values that need replacing with a newer one, we insert the new one
+  # to the array (inserted 'in place' to retain original order). Then remove
+  # the old facet values in bulk and write only the unique values that remain.
+  def update_facet_values(subscriber_lists)
+    subscriber_lists.each do |list|
+      links = list.links[:facet_values][:any]
+      updated_links = links.dup # copy the links to avoid corrupting the loop
+      links.each_with_index do |value, index|
+        replacement = EuExitFacetMigrationConfig::facet_values_to_replace.find { |replace_val| replace_val[:old_facet_value] == value }
+        updated_links.insert(index, replacement[:new_facet_value]) if replacement
+      end
+      updated_links = updated_links - EuExitFacetMigrationConfig::facet_values_to_remove
+      updated_links = updated_links.uniq
+      list.write_attribute('links', facet_values: { any: updated_links })
+      list.write_attribute('title', title_from_facets(updated_links))
+      list.save
+    end
+  end
+
+  # Title is constructed from finder content item, so we have to grab the JSON.
+  def title_from_facets(links)
```

Comment:
Is there any possibility to use `EmailAlertListTitleBuilder` instead of copying much of the code into this rake task?  If this rake task ever needs to be used again, it would save future developers having to check the code and hard-coded text matches whatever is in `EmailAlertListTitleBuilder`.

Reply:
I agree, this was my first instinct & I copied EmailAlertListTitleBuilder to try to use that, but found I was doing so much data manipulation to get the content item JSON in the structure EmailAlertListTitleBuilder that it was a better decision to roll out a pared down version myself.

The code that generates the email title from finder-frontend conflates three things: all the possible facet values, and the user's facet value choices, and the groupings of the facet values inside different facets. So we'd have a structure like:

```json
filter: {
  "sector_business_area" => [
    "content-id-of-a-facet-the-user-selected",
    "content-id-of-another-facet-the-user-selected"
  ],
  "buying_selling_in_the_eu" => [
    "yet-another",
  ] // etc
}
```

The facet groupings were required for the email alert frontend but what it all comes down to in email subscriber list is keys for the chosen facet-values:

```
[
  "content-id-of-a-facet-the-user-selected",
  "content-id-of-another-facet-the-user-selected"
  "yet-another",
]
```

So the solution for the rake task is simplest when we just take that array of content IDs and find the corresponding `title`s in the content item, rather than try and map it to its facet group (which has no bearing on the outcome here) just to shoehorn it into a local copy of EmailAlertListTitleBuilder.

Reply:
I've now added this explanation into the commit message (and fixed up the linting issues).

Still hoping to refactor the `title_from_facets` method before it's truly 'review ready'.